webui/react/src/hooks/useModal/UserSettings/useModalCreateUser.tsx

```diff
@@ -196,9 +230,19 @@ const useModalCreateUser = ({ groups, onClose, user }: ModalProps): ModalHooks =
       await form.validateFields();
 
       const formData = form.getFieldsValue();
+
+      const newRoles: Set<number> = new Set(formData.roles);
+      // TODO user should not be able to set roles before they load.
+      const oldRoles = new Set((userRoles ?? []).map((r) => r.id));
+
+      const rolesToAdd = filter((r: number) => !oldRoles.has(r))(newRoles);
+      const rolesToRemove = filter((r: number) => !newRoles.has(r))(oldRoles);
+
       try {
         if (user) {
           await patchUser({ userId: user.id, userParams: formData });
+          await assignRolesToUser({ roleIds: Array.from(rolesToAdd), userId: user.id });
+          await removeRolesFromUser({ roleIds: Array.from(rolesToRemove), userId: user.id });
```

Comment:
I tried to also update the local state here so that it will be reflected if the user re-opens the modal but it seems there is some caching going on with the antd form. Is there some kind of form reset function that's supposed to be called after a submit?

Reply:
`form.resetFields()` should work